byterun/Makefile

```diff
@@ -207,28 +206,39 @@ libcamlrun_pic.$(A): $(PICOBJS)
 libcamlrun_shared.$(SO): $(PICOBJS)
 	$(MKDLL) -o $@ $^ $(BYTECCLIBS)
 
-%.$(O): %.c
-	$(CC) -c $(CFLAGS) $(CPPFLAGS) $<
+# The COMPILE_C_FILE function below receives as argument the pattern
+# that corresponds to the name of the generated object file
+define COMPILE_C_FILE
+$(1): %.c
+	$$(CC) -c $$(OC_CFLAGS) $$(OC_CPPFLAGS) $$(OUTPUTOBJ)$$@ $$<
+endef
 
-%.$(DBGO): %.c
-	$(CC) -c $(DFLAGS) $(CPPFLAGS) $(OUTPUTOBJ)$@ $<
+object_types := %.$(O) %.d.$(O) %.i.$(O) %.pic.$(O)
+
+$(foreach object_type, $(object_types), \
+  $(eval $(call COMPILE_C_FILE,$(object_type))))
+
+%.$(DBGO): OC_CPPFLAGS += $(OC_DEBUG_CPPFLAGS)
+
+%.i.$(O): OC_CPPFLAGS += $(OC_INSTR_CPPFLAGS)
+
+%.pic.$(O): OC_CFLAGS += $(SHAREDLIB_CFLAGS)
```

Comment:
It feels like this should be shared between `asmrun` and `byterun`, rather than duplicated? Given that we're getting into the realm of GNU make macros!